final_project/perturb.py

```diff
@@ -0,0 +1,298 @@
+"""
+APC 524 
+Final Project
+Quantifying resilience of coupled networks
+
+@author Jalal Butt
+Date: 12-03-22
+
+
+Module file handling the perturbation network
+----------------------------------------------
+
+
+Inputs:
+-------
+	- grid obj 
+
+
+Outputs:
+--------
+	- time-evolution of perturbation
+		--> entirely independent of what's happening on the grid. I.e., need not output 
+			individual time-steps
+"""
+
+
+
+## (Elliptic Equations hw)
+import numpy as np
+import matplotlib.pyplot as plt
+import scipy.sparse as sparse
+import scipy.sparse.linalg as linalg
+# from scipy import signal
+
+from dataclasses import dataclass
+
+
+
+## external inputs, to be fed in from 
+m = 100 # number of grid-points
+# n = 100 #  currently not implemented
+L_x = 10
+H_y = 10
+
+# f_type = 'oscillatory'
+f_type = 'delta'
+source_point = (-2,-2)
+source_strength = 1000
+
+plotSolution = True
+plotError = False
+
+
+
+
+# Problem specification
+x = np.linspace(0, L_x, m + 2)
+y = np.linspace(0, H_y, m + 2)
+delta_x = L_x / (m + 1)
+
+
+@dataclass
+class PerturbedNetwork:
+	"""
+	Class for the perturbation network model. 
+
+	Inputs:
+	------
+		m
+		...
+
+	Methods:
+	--------
+		construct_A:   construct the spatial operator
+		construct_source:   construct perturb source (several options)
+		solve_laplacian_system:   numerically solve laplacian
+		compute_discretization_error:   <only valid for an analytic RHS>
+		plot_solution:   plot the solution
+		plot_error:    plot the error
+
+		static_solve:   compute the time-independent solution
+
+
+
+
+	"""
+	def __init__(self, f_type, source_point, L_x, H_y, m, source_strength= 1000):
+		self.f_type          = f_type
+		self.source_point    = source_point
+		self.source_strength = source_strength
+		self.m   = m
+		self.L_x = L_x
+		self.H_y = H_y
+
+
+		self.x = np.linspace(0, L_x, m + 2)
+		self.y = np.linspace(0, H_y, m + 2)
+		self.delta_x = L_x / (m + 1)		
+
+
+
+
+
+	def __post_init__():  # what should i do w this? generate tests?
```

Comment:
this is for if you need to do any validations, or other initializations other than data